google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -608,8 +613,9 @@ public boolean shouldRetryListener(BulkWriterError error) {
         ApiFutures.transformAsync(
             executeWriteHelper(documentReference, operationType, operationCallback, 0),
             new ApiAsyncFunction<WriteResult, WriteResult>() {
-              public ApiFuture<WriteResult> apply(WriteResult result) {
-                successListener.onResult(documentReference, result);
+              public ApiFuture<WriteResult> apply(WriteResult result)
+                  throws ExecutionException, InterruptedException {
+                invokeUserSuccessCallback(documentReference, result).get();
```

Comment:
This looks like it may block the GRPC thread while the user code is running. Instead, you should do one of the following:

- Run your code
- Run user code in user thread, followed by a continuation that dispatches back on your old exectur
- Finish running your own code

Realistically, though, no one is going to get mad if you run a little bit of your own code on the use executor (as long as it is non-blocking and doesn't throw). For that reason, you can likely just use the user executor when the callback is first invoked (e.g. replace the directExecutor call in `executeWrite()` if deemed safe).

Reply:
Using userExecutor for `executeWrite()`, and got rid of the invokeCallback methods.

Reply:
Summary of conversation w/ Ben: We should not block grpc or gax threads. Instead, we should be creating a thread on each firestore instance that awaits user callbacks.

Solution: Created a separate thread for waiting on the callback. Fixed async issue by fixing bug in `performFlush()`, where we were incorrectly resolving the `flushComplete` future early if no retries were scheduled.